issues with neovim 0.5.0 on an M1 macbook

**Describe the bug**

Since #110837 made it into nixpkgs-unstable I can't install neovim on an M1 laptop anymore, error is:

```
$ nix-env -iA nixpkgs.neovim
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
installing 'neovim-0.5.0'
these derivations will be built:
  /nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv
  /nix/store/4f1kkp8d9gqzv2ki8y1ndb68n3r92qi4-luajit-2.1.0-2021-05-29-luv-1.30.0-0.drv
  /nix/store/7ixqv6aqw79g124nzlm2nfz85xl7n3lm-luajit-2.1.0-2021-05-29-lpeg-1.0.2-1.drv
  /nix/store/h42jqwavkx330ibqz3216zpa3n79j5sc-luajit-2.1.0-2021-05-29-luabitop-1.0.2-3.drv
  /nix/store/lkw5165jvhhn6v4vl3m6dysy75jasdwb-luajit-2.1.0-2021-05-29-mpack-1.0.8-0.drv
  /nix/store/jd1p97vxrfh56vnb8l2di4az3wij90z1-luajit-2.1.0-2021-05-29-env.drv
  /nix/store/wz7l2p5qkav720fs6anl766gdybc6ppd-neovim-unwrapped-0.5.0.drv
  /nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv
building '/nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv'...
unpacking sources
unpacking source archive /nix/store/33jwc6ncf287ziykl5nd5y01gdb5di89-source
source root is source
patching sources
applying patch /nix/store/m4lqm4f5zvpbqv4hpawz8cvl0snrldhl-darwin-3.1.3.patch
patching file src/luarocks/core/cfg.lua
Hunk #1 succeeded at 427 (offset 2 lines).
Hunk #2 succeeded at 436 (offset 2 lines).
updateAutotoolsGnuConfigScriptsPhase
configuring
configure flags: --prefix=/nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 --with-lua-include=/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1/

Configuring LuaRocks version 3.2.1...

Lua version detected: 5.1
Lua interpreter found: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/bin/luajit
lua.h found: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1/lua.h
unzip found in PATH: /nix/store/2izly43wxa0h3jvd8rzfkc7q1rg74b2j-unzip-6.0/bin

Done configuring.

LuaRocks will be installed at......: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1
LuaRocks will install rocks at.....: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1
LuaRocks configuration directory...: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1/etc/luarocks
Using Lua from.....................: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29
Lua include directory..............: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1

* Type make and make install:
  to install to /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 as usual.
* Type make bootstrap:
  to install LuaRocks into /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 as a rock.

building
build flags: SHELL=/nix/store/zj3mf2cki8hcs5qmkdns85nkncdqsmjb-bash-4.4-p23/bin/bash
mkdir -p "build"
mkdir -p .luarocks
cp ./build/config-5.1.lua .luarocks/config-5.1.lua
rm -f src/luarocks/core/hardcoded.lua
echo "#!/bin/sh" > luarocks
echo "unset LUA_PATH LUA_PATH_5_2 LUA_PATH_5_3 LUA_PATH_5_4 LUA_CPATH LUA_CPATH_5_2 LUA_CPATH_5_3 LUA_CPATH_5_4" >> luarocks
echo 'LUAROCKS_SYSCONFDIR="/nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1/etc/luarocks" LUA_PATH="/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/?.lua;;" exec "/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/bin/luajit" "/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/bin/luarocks" --project-tree="/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/lua_modules" "$@"' >> luarocks
chmod +rx ./luarocks
./luarocks init
PANIC: unprotected error in call to Lua API (module 'luarocks.core.hardcoded' not found:
        no field package.preload['luarocks.core.hardcoded']
        no file '/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/luarocks/core/hardcoded.lua'
        no file './luarocks/core/hardcoded.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/luajit-2.1.0-beta3/luarocks/core/hardcoded.lua'
        no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded.lua'
        no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded/init.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded/init.lua'
        no file './luarocks/core/hardcoded.so'
        no file '/usr/local/lib/lua/5.1/luarocks/core/hardcoded.so'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks/core/hardcoded.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './luarocks.so'
        no file '/usr/local/lib/lua/5.1/luarocks.so'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks.so'
        no file '/usr/local/lib/lua/5.1/loadall.so')
make: *** [GNUmakefile:57: luarocks] Error 1
builder for '/nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv' failed with exit code 2
cannot build derivation '/nix/store/7ixqv6aqw79g124nzlm2nfz85xl7n3lm-luajit-2.1.0-2021-05-29-lpeg-1.0.2-1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/h42jqwavkx330ibqz3216zpa3n79j5sc-luajit-2.1.0-2021-05-29-luabitop-1.0.2-3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/4f1kkp8d9gqzv2ki8y1ndb68n3r92qi4-luajit-2.1.0-2021-05-29-luv-1.30.0-0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/lkw5165jvhhn6v4vl3m6dysy75jasdwb-luajit-2.1.0-2021-05-29-mpack-1.0.8-0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wz7l2p5qkav720fs6anl766gdybc6ppd-neovim-unwrapped-0.5.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv' failed
➜  dotfiles git:(master) ✗
➜  dotfiles git:(master) ✗ curl -L https://nixos.org/nix/install | sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  2699  100  2699    0     0   3160      0 --:--:-- --:--:-- --:--:--  3160
downloading Nix 2.3.13 binary tarball for x86_64-darwin from 'https://releases.nixos.org/nix/nix-2.3.13/nix-2.3.13-x86_64-darwin.tar.xz' to '/var/folders/8y/1jwycq1x595d0_3g887dbjm40000gn/T/nix-binary-tarball-unpack.XXXXXXXXXX.tuAW1oi6'...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 34.6M  100 34.6M    0     0  7281k      0  0:00:04  0:00:04 --:--:-- 7772k
^Csh: SHA-256 hash mismatch in 'https://releases.nixos.org/nix/nix-2.3.13/nix-2.3.13-x86_64-darwin.tar.xz'; expected 81cc3f2d5b06162b43a4695b37952ef569318d44c999558b8c54a925bab24ded, got
➜  dotfiles git:(master) ✗ nix-env -iA nixpkgs.neovim
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
installing 'neovim-0.5.0'
these derivations will be built:
  /nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv
  /nix/store/4f1kkp8d9gqzv2ki8y1ndb68n3r92qi4-luajit-2.1.0-2021-05-29-luv-1.30.0-0.drv
  /nix/store/7ixqv6aqw79g124nzlm2nfz85xl7n3lm-luajit-2.1.0-2021-05-29-lpeg-1.0.2-1.drv
  /nix/store/h42jqwavkx330ibqz3216zpa3n79j5sc-luajit-2.1.0-2021-05-29-luabitop-1.0.2-3.drv
  /nix/store/lkw5165jvhhn6v4vl3m6dysy75jasdwb-luajit-2.1.0-2021-05-29-mpack-1.0.8-0.drv
  /nix/store/jd1p97vxrfh56vnb8l2di4az3wij90z1-luajit-2.1.0-2021-05-29-env.drv
  /nix/store/wz7l2p5qkav720fs6anl766gdybc6ppd-neovim-unwrapped-0.5.0.drv
  /nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv
building '/nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv'...
unpacking sources
unpacking source archive /nix/store/33jwc6ncf287ziykl5nd5y01gdb5di89-source
source root is source
patching sources
applying patch /nix/store/m4lqm4f5zvpbqv4hpawz8cvl0snrldhl-darwin-3.1.3.patch
patching file src/luarocks/core/cfg.lua
Hunk #1 succeeded at 427 (offset 2 lines).
Hunk #2 succeeded at 436 (offset 2 lines).
updateAutotoolsGnuConfigScriptsPhase
configuring
configure flags: --prefix=/nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 --with-lua-include=/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1/

Configuring LuaRocks version 3.2.1...

Lua version detected: 5.1
Lua interpreter found: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/bin/luajit
lua.h found: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1/lua.h
unzip found in PATH: /nix/store/2izly43wxa0h3jvd8rzfkc7q1rg74b2j-unzip-6.0/bin

Done configuring.

LuaRocks will be installed at......: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1
LuaRocks will install rocks at.....: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1
LuaRocks configuration directory...: /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1/etc/luarocks
Using Lua from.....................: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29
Lua include directory..............: /nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/include/luajit-2.1

* Type make and make install:
  to install to /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 as usual.
* Type make bootstrap:
  to install LuaRocks into /nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1 as a rock.

building
build flags: SHELL=/nix/store/zj3mf2cki8hcs5qmkdns85nkncdqsmjb-bash-4.4-p23/bin/bash
mkdir -p "build"
mkdir -p .luarocks
cp ./build/config-5.1.lua .luarocks/config-5.1.lua
rm -f src/luarocks/core/hardcoded.lua
echo "#!/bin/sh" > luarocks
echo "unset LUA_PATH LUA_PATH_5_2 LUA_PATH_5_3 LUA_PATH_5_4 LUA_CPATH LUA_CPATH_5_2 LUA_CPATH_5_3 LUA_CPATH_5_4" >> luarocks
echo 'LUAROCKS_SYSCONFDIR="/nix/store/cmfi3bavcszmj7ld9jg6hnzn3ky0a6wa-luarocks-3.2.1/etc/luarocks" LUA_PATH="/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/?.lua;;" exec "/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/bin/luajit" "/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/bin/luarocks" --project-tree="/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/lua_modules" "$@"' >> luarocks
chmod +rx ./luarocks
./luarocks init
PANIC: unprotected error in call to Lua API (module 'luarocks.core.hardcoded' not found:
        no field package.preload['luarocks.core.hardcoded']
        no file '/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/luarocks/core/hardcoded.lua'
        no file './luarocks/core/hardcoded.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/luajit-2.1.0-beta3/luarocks/core/hardcoded.lua'
        no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded.lua'
        no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded/init.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded.lua'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded/init.lua'
        no file './luarocks/core/hardcoded.so'
        no file '/usr/local/lib/lua/5.1/luarocks/core/hardcoded.so'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks/core/hardcoded.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
        no file './luarocks.so'
        no file '/usr/local/lib/lua/5.1/luarocks.so'
        no file '/nix/store/svpw0x20dx1sswf6imfa2xjfb7j4rjj7-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks.so'
        no file '/usr/local/lib/lua/5.1/loadall.so')
make: *** [GNUmakefile:57: luarocks] Error 1
builder for '/nix/store/5gp4g1718m3d8h2c1i07iaab557qhjb3-luarocks-3.2.1.drv' failed with exit code 2
cannot build derivation '/nix/store/7ixqv6aqw79g124nzlm2nfz85xl7n3lm-luajit-2.1.0-2021-05-29-lpeg-1.0.2-1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/h42jqwavkx330ibqz3216zpa3n79j5sc-luajit-2.1.0-2021-05-29-luabitop-1.0.2-3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/4f1kkp8d9gqzv2ki8y1ndb68n3r92qi4-luajit-2.1.0-2021-05-29-luv-1.30.0-0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/lkw5165jvhhn6v4vl3m6dysy75jasdwb-luajit-2.1.0-2021-05-29-mpack-1.0.8-0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wz7l2p5qkav720fs6anl766gdybc6ppd-neovim-unwrapped-0.5.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/m200hhv91ba5y9j3sgf1vyb1spa30x1x-neovim-0.5.0.drv' failed
```


**To Reproduce**
Steps to reproduce the behavior:
1. On an M1 macbook run `nix-env -iA nixpkgs.neovim`

**Expected behavior**
neovim to successfully install

**Screenshots**
N/A

**Additional context**


**Notify maintainers**
@rvolosatovs 

**Metadata**

```
$ nix-shell -p nix-info --run "nix-info -m"
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
these paths will be fetched (0.32 MiB download, 1.90 MiB unpacked):
  /nix/store/0ljw5ppdzw17p2rl2j3fbf28a5pqa0xa-bash-interactive-4.4-p23-doc
  /nix/store/55cs339a40h0qdk2m6hjg7hyxcws4l55-bash-interactive-4.4-p23-dev
copying path '/nix/store/0ljw5ppdzw17p2rl2j3fbf28a5pqa0xa-bash-interactive-4.4-p23-doc' from 'https://cache.nixos.org'...
copying path '/nix/store/55cs339a40h0qdk2m6hjg7hyxcws4l55-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
 - system: `"aarch64-darwin"`
 - host os: `Darwin 20.5.0, macOS 11.4`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.12`
 - channels(dbarroso): `"darwin, home-manager, nixpkgs-21.11pre299898.c40d0126f85"`
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
 - nixpkgs: `/Users/dbarroso/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -27487,7 +27487,32 @@ in
   wrapNeovimUnstable = callPackage ../applications/editors/neovim/wrapper.nix { };
   wrapNeovim = neovim-unwrapped: lib.makeOverridable (neovimUtils.legacyWrapper neovim-unwrapped);
   neovim-unwrapped = callPackage ../applications/editors/neovim {
-    lua = luajit;
+    # neovim doesn't build with luajit on aarch64:
+    # ./luarocks init
+    # PANIC: unprotected error in call to Lua API (module 'luarocks.core.hardcoded' not found:
+    #         no field package.preload['luarocks.core.hardcoded']
+    #         no file '/private/tmp/nix-build-luarocks-3.2.1.drv-0/source/src/luarocks/core/hardcoded.lua'
+    #         no file './luarocks/core/hardcoded.lua'
+    #         no file '/nix/store/3s6c509q9vvq3db87rfi7qa38wzxwz8w-luajit-2.1.0-2021-05-29/share/luajit-2.1.0-beta3/luarocks/core/hardcoded.lua'
+    #         no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded.lua'
+    #         no file '/usr/local/share/lua/5.1/luarocks/core/hardcoded/init.lua'
+    #         no file '/nix/store/3s6c509q9vvq3db87rfi7qa38wzxwz8w-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded.lua'
+    #         no file '/nix/store/3s6c509q9vvq3db87rfi7qa38wzxwz8w-luajit-2.1.0-2021-05-29/share/lua/5.1/luarocks/core/hardcoded/init.lua'
+    #         no file './luarocks/core/hardcoded.so'
+    #         no file '/usr/local/lib/lua/5.1/luarocks/core/hardcoded.so'
+    #         no file '/nix/store/3s6c509q9vvq3db87rfi7qa38wzxwz8w-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks/core/hardcoded.so'
+    #         no file '/usr/local/lib/lua/5.1/loadall.so'
+    #         no file './luarocks.so'
+    #         no file '/usr/local/lib/lua/5.1/luarocks.so'
+    #         no file '/nix/store/3s6c509q9vvq3db87rfi7qa38wzxwz8w-luajit-2.1.0-2021-05-29/lib/lua/5.1/luarocks.so'
+    #         no file '/usr/local/lib/lua/5.1/loadall.so')
+    # make: *** [GNUmakefile:57: luarocks] Error 1
+    #
+    # See https://github.com/NixOS/nixpkgs/issues/129099
+    # Possibly related: https://github.com/neovim/neovim/issues/7879
+    lua =
+      if stdenv.isAarch64 then lua5_1 else
+      luajit;
   };
 
   neovimUtils = callPackage ../applications/editors/neovim/utils.nix { };

